sentry_sdk/metrics.py

```diff
@@ -0,0 +1,575 @@
+    # type: (str, Optional[MetricTags]) -> Tuple[Optional[MetricsAggregator], Optional[MetricTags]]
+    """Returns the current metrics aggregator if there is one."""
+    hub = Hub.current
+    client = hub.client
+    if client is None or client.metrics_aggregator is None:
+        return None, tags
+
+    updated_tags = dict(tags or ())  # type: Dict[str, MetricTagValue]
+    updated_tags.setdefault("release", client.options["release"])
+    updated_tags.setdefault("environment", client.options["environment"])
+
+    scope = hub.scope
+    transaction_source = scope._transaction_info.get("source")
+    if transaction_source in GOOD_TRANSACTION_SOURCES:
+        transaction = scope._transaction
+        if transaction:
+            updated_tags.setdefault("transaction", transaction)
+
+    callback = client.options.get("_experiments", {}).get("before_emit_metric")
+    if callback is not None:
+        if not callback(key, updated_tags):
+            return None, updated_tags
+
+    return client.metrics_aggregator, updated_tags
+
+
+def incr(
+    key,  # type: str
+    value=1.0,  # type: float
+    unit="none",  # type: MetricUnit
+    tags=None,  # type: Optional[MetricTags]
+    timestamp=None,  # type: Optional[float]
+):
+    # type: (...) -> None
+    """Increments a counter."""
+    aggregator, tags = _get_aggregator_and_update_tags(key, tags)
+    if aggregator is not None:
+        aggregator.add("c", key, value, unit, tags, timestamp)
+
+
+class _Timing(object):
+    def __init__(
+        self,
+        key,  # type: str
+        tags,  # type: Optional[MetricTags]
+        timestamp,  # type: Optional[float]
+    ):
+        # type: (...) -> None
+        self.key = key
+        self.tags = tags
+        self.timestamp = timestamp
+        self.then = None  # type: Optional[float]
+
+    def __enter__(self):
+        # type: (...) -> _Timing
+        self.then = now()
+        return self
+
+    def __exit__(self, exc_type, exc_value, tb):
+        # type: (Any, Any, Any) -> None
+        aggregator, tags = _get_aggregator_and_update_tags(self.key, self.tags)
+        if aggregator is not None:
+            elapsed = now() - self.then  # type: ignore
+            aggregator.add("d", self.key, elapsed, "second", tags, self.timestamp)
+
+    def __call__(self, f):
+        # type: (Any) -> Any
+        @wraps(f)
+        def timed_func(*args, **kwargs):
+            # type: (*Any, **Any) -> Any
+            with timing(self.key, self.tags, self.timestamp):
+                return f(*args, **kwargs)
+
+        return timed_func
+
+
+def timing(
+    key,  # type: str
+    tags=None,  # type: Optional[MetricTags]
+    timestamp=None,  # type: Optional[float]
+):
+    # type: (...) -> _Timing
+    """Emits a distribution with the time it takes to run the given code block."""
+    return _Timing(key, tags, timestamp)
+
+
+def distribution(
+    key,  # type: str
+    value,  # type: float
+    unit="second",  # type: MetricUnit
```

Comment:
Thought: What if you make the default unit of `distribution=None` and instead have a non-contextmanager version of `timing` that has `second` (or `millisecond` or whatever)? On a function with such a general name, it's not necessarily clear that the default unit should be seconds and it can be easily missed.

Reply:
I'm a bit conflicted on this. Most distributions are likely to be timers and I'm assuming that people will miss out on the metric.

Reply:
Resolved now.